error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,143 @@
+ * with the original code should be revisited and fixed in a structural manner by the developer.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "`Mono#zip` and `Mono#zipWith` should not be executed against `Mono#empty` or `Mono<Void>` parameter; "
+            + "please revisit the parameters used and make sure to supply correct publishers instead",
+    link = BUG_PATTERNS_BASE_URL + "MonoZipOfMonoVoidUsage",
+    linkType = CUSTOM,
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO = type("reactor.core.publisher.Mono");
+  // In fact, we use `Mono<Void>` everywhere in codebases instead of `Mono<Object>` to represent
+  // empty publisher
+  private static final Supplier<Type> MONO_VOID_TYPE =
+      VisitorState.memoize(generic(MONO, type("java.lang.Void")));
+
+  // On instance mono.zipWith, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> MONO_ZIP_AND_WITH =
+      anyOf(
+          allOf(
+              instanceMethod().onDescendantOf(MONO).named("zipWith"),
+              toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE))),
+          allOf(
+              instanceMethod().onDescendantOf(MONO).named("zipWith"),
+              toType(MethodInvocationTree.class, staticMethod().onClass(MONO).named("empty"))),
+          onClassWithMethodName(MONO_VOID_TYPE, "zipWith"));
+
+  // On class Mono.zip, at least one element should match empty in order to proceed.
+  private static final Matcher<ExpressionTree> STATIC_MONO_ZIP =
+      allOf(
+          staticMethod().onClass(MONO).named("zip"),
+          toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE)));
+
+  /** Instantiates a new {@link MonoZipOfMonoVoidUsage} instance. */
+  public MonoZipOfMonoVoidUsage() {}
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+    boolean dynamicMono = MONO_ZIP_AND_WITH.matches(tree, state);
+    boolean staticMono = STATIC_MONO_ZIP.matches(tree, state);
+    if (!dynamicMono && !staticMono) {
+      return Description.NO_MATCH;
+    }
+
+    return buildDescription(tree)
+        .setMessage(
+            "`Mono#zip` and `Mono#zipWith` should not be executed against empty publisher; "
+                + "remove it or suppress this warning and add a comment explaining its purpose")
+        .addFix(SuggestedFixes.addSuppressWarnings(state, canonicalName()))
+        .build();
+  }
+
+  private static Matcher<ExpressionTree> onClassWithMethodName(
+      Supplier<Type> genericDesiredType, String methodName) {
+    return (tree, state) -> {
+      JCTree.JCExpression methodSelect = ((JCTree.JCMethodInvocation) tree).getMethodSelect();
```

Comment:
```suggestion
 MemberSelectTree ms = (MemberSelectTree) tree.getMethodSelect();
```
Something like that? Didn't check out so I'm not sure.

If we use that, the rest of the method probably looks a bit nicer.

Reply:
I've double checked. We need the real implementation class, not the abstract interface, otherwise I don't have access to the needed fields under the hood

Reply:
👌🏼